README.md
# karma-junit-reporter

[![js-standard-style](https://img.shields.io/badge/code%20style-standard-brightgreen.svg?style=flat-square)](https://github.com/karma-runner/karma-junit-reporter)
 [![npm version](https://img.shields.io/npm/v/karma-junit-reporter.svg?style=flat-square)](https://www.npmjs.com/package/karma-junit-reporter) [![npm downloads](https://img.shields.io/npm/dm/karma-junit-reporter.svg?style=flat-square)](https://www.npmjs.com/package/karma-junit-reporter)

[![Build Status](https://img.shields.io/travis/karma-runner/karma-junit-reporter/master.svg?style=flat-square)](https://travis-ci.org/karma-runner/karma-junit-reporter) [![Dependency Status](https://img.shields.io/david/karma-runner/karma-junit-reporter.svg?style=flat-square)](https://david-dm.org/karma-runner/karma-junit-reporter) [![devDependency Status](https://img.shields.io/david/dev/karma-runner/karma-junit-reporter.svg?style=flat-square)](https://david-dm.org/karma-runner/karma-junit-reporter#info=devDependencies)

> Reporter for the JUnit XML format.

## Installation

The easiest way is to keep `karma-junit-reporter` as a devDependency in your `package.json`. Just run

```bash
npm install karma-junit-reporter --save-dev
```

to let npm automatically add it there.

## Configuration

```js
// karma.conf.js
module.exports = function(config) {
  config.set({
    reporters: ['progress', 'junit'],

    // the default configuration
    junitReporter: {
      outputDir: '', // results will be saved as $outputDir/$browserName.xml
      outputFile: undefined, // if included, results will be saved as $outputDir/$browserName/$outputFile
      suite: '', // suite will become the package name attribute in xml testsuite element
      useBrowserName: true, // add browser name to report and classes names
      nameFormatter: undefined, // function (browser, result) to customize the name attribute in xml testcase element
      classNameFormatter: undefined, // function (browser, result) to customize the classname attribute in xml testcase element
      properties: {}, // key value pair of properties to add to the <properties> section of the report
      xmlVersion: null, // use '1' if reporting to be per SonarQube 6.2 XML format
      gitlab: null, // use 'true' if you want to use Gitlab parsers
    }
  });
};
```

You can pass list of reporters as a CLI argument too:
```bash
karma start --reporters junit,dots
```

## Produce test result with schema acceptable in sonar

To make this possible, it's required to make the classnames of each tests to match its file name.

For Example:
```js
describe('analytics.AnalyticsModule_test', function(){

    var analytics;
    beforeEach(module('ECApp'));
    beforeEach(module('angularytics'));
    beforeEach(module('AnalyticsModule'));
...
```

should have a file name AnalyticsModule_test.js

This will produce test result with schema acceptable in sonar.

Grunt file reporters property example:
```js
reporters: ['junit', 'coverage', 'progress'],
junitReporter: {
    outputDir: $junitResults,
    suite: 'models'
},
coverageReporter: {
    type: 'lcov',
    dir: $coverageOutputDir,
    subdir: '.'
},
preprocessors: {
    'src/main/webapp/public/js/ec3.3/**/*.js': 'coverage',
    'src/main/webapp/public/js/ec3/**/*.js': 'coverage'
},
plugins: [
    'karma-jasmine',
    'karma-phantomjs-launcher',
    'ec-karma-junit-reporter23',
    'karma-coverage'
]
```

Sonar property example:
```js
sonar.projectName=js
sonar.sources=site-main-php/src/main/webapp/public/js
sonar.projectBaseDir=.
sonar.exclusions=site-main-php/src/main/webapp/public/js/lib/*.js,site-main-php/src/main/webapp/public/js/tests/**/*.php,site-main-php/src/main/webapp/public/js/tests/**/*.js,site-main-php/src/main/webapp/public/js/ec3.3/vendor/**
sonar.javascript.lcov.reportPath=site-main-php/target/coverage/lcov.info
sonar.javascript.jstestdriver.reportsPath=site-main-php/target/surefire-reports/
sonar.tests=site-main-php/src/main/webapp/public/js/tests
```

Example junit xml report:
```xml
<?xml version="1.0"?>
<testsuite name="PhantomJS 1.9.8 (Linux)" package="models" timestamp="2015-03-10T13:59:23" id="0" hostname="admin" tests="629" errors="0" failures="0" time="11.452">
  <properties>
    <property name="browser.fullName" value="Mozilla/5.0 (Unknown; Linux x86_64) AppleWebKit/534.34 (KHTML, like Gecko) PhantomJS/1.9.8 Safari/534.34"/>
  </properties>
 <testcase name="(C.2) Checks if an empty object is returned when error 404 is encountered" time="0.01" classname="PhantomJS_1_9_8_(Linux).models.AnalyticsModule_test"/>
 <testcase name="(C.3) Checks if an empty array is returned when error 405 is encountered" time="0.013" classname="PhantomJS_1_9_8_(Linux).models.AnalyticsModule_test"/>
</testsuite>
...
```
----

For more information on Karma see the [homepage].


[homepage]: http://karma-runner.github.com



CHANGELOG.md
<a name="2.0.0"></a>
# [2.0.0](https://github.com/karma-runner/karma-junit-reporter/compare/v1.2.0...v2.0.0) (2019-10-07)


### Bug Fixes

* allow special characters and emoji in output ([59fd3c6](https://github.com/karma-runner/karma-junit-reporter/commit/59fd3c6))


### Chores

* **travis:** Drop old nodejs version: NO v10 ([#164](https://github.com/karma-runner/karma-junit-reporter/issues/164)) ([bc07177](https://github.com/karma-runner/karma-junit-reporter/commit/bc07177))
* remove support for node 6 ([435cb5e](https://github.com/karma-runner/karma-junit-reporter/commit/435cb5e))


### Features

* support for new XML format (per SonarQube)  ([b0984e0](https://github.com/karma-runner/karma-junit-reporter/commit/b0984e0))


### BREAKING CHANGES

* Drop Support for Node 6, to make it possible to use async/await in karma codebase.
* **travis:** no support for node < 6

* node 10 not supported by libxmljs-mt



<a name="1.2.0"></a>
# [1.2.0](https://github.com/karma-runner/karma-junit-reporter/compare/v1.1.0...v1.2.0) (2016-12-06)


### Bug Fixes

* initialize var suites early  ([e09acb2](https://github.com/karma-runner/karma-junit-reporter/commit/e09acb2))
* release memory held by a testsuite after we're done with it. ([eacf8bb](https://github.com/karma-runner/karma-junit-reporter/commit/eacf8bb))
* remove unnecessary nulling out of `suites` ([4202ee8](https://github.com/karma-runner/karma-junit-reporter/commit/4202ee8)), closes [#99](https://github.com/karma-runner/karma-junit-reporter/issues/99)



<a name="1.1.0"></a>
# [1.1.0](https://github.com/karma-runner/karma-junit-reporter/compare/v1.0.0...v1.1.0) (2016-06-26)


### Bug Fixes

* add defensive checks to safely handle browser disconnects. ([485d87a](https://github.com/karma-runner/karma-junit-reporter/commit/485d87a))



<a name="1.0.0"></a>
# [1.0.0](https://github.com/karma-runner/karma-junit-reporter/compare/v0.4.2...v1.0.0) (2016-05-03)




<a name="0.4.2"></a>
## [0.4.2](https://github.com/karma-runner/karma-junit-reporter/compare/v0.4.1...v0.4.2) (2016-04-08)


### Features

* Add support for additional properties to add to the report section ([b67d234](https://github.com/karma-runner/karma-junit-reporter/commit/b67d234))



<a name="0.4.1"></a>
## [0.4.1](https://github.com/karma-runner/karma-junit-reporter/compare/v0.4.0...v0.4.1) (2016-03-21)


### Bug Fixes

* check if outputFile is set before checking for absolute outputFile ([c8887ac](https://github.com/karma-runner/karma-junit-reporter/commit/c8887ac)), closes [#87](https://github.com/karma-runner/karma-junit-reporter/issues/87)



<a name="0.4.0"></a>
# [0.4.0](https://github.com/karma-runner/karma-junit-reporter/compare/v0.3.8...v0.4.0) (2016-03-10)


### Bug Fixes

* Handle absolute outputFile paths in Junit Reporter ([d5dc808](https://github.com/karma-runner/karma-junit-reporter/commit/d5dc808)), closes [#83](https://github.com/karma-runner/karma-junit-reporter/issues/83)
* test name to contain parent suite(s) names ([8e6e202](https://github.com/karma-runner/karma-junit-reporter/commit/8e6e202)), closes [#62](https://github.com/karma-runner/karma-junit-reporter/issues/62)

### Features

* add name and classname formatters ([3f43c51](https://github.com/karma-runner/karma-junit-reporter/commit/3f43c51)), closes [#75](https://github.com/karma-runner/karma-junit-reporter/issues/75)



<a name="0.3.8"></a>
## [0.3.8](https://github.com/karma-runner/karma-junit-reporter/compare/v0.3.7...v0.3.8) (2015-10-20)


### Features

* **reporter:** add useBrowserName parameter ([2327234](https://github.com/karma-runner/karma-junit-reporter/commit/2327234))



<a name="0.3.7"></a>
## [0.3.7](https://github.com/karma-runner/karma-junit-reporter/compare/v0.3.6...v0.3.7) (2015-10-01)


### Bug Fixes

* **reporter:** bug creating directory when outputFile specified (#67) ([2205f47](https://github.com/karma-runner/karma-junit-reporter/commit/2205f47))



<a name="0.3.6"></a>
## [0.3.6](https://github.com/karma-runner/karma-junit-reporter/compare/v0.3.5...v0.3.6) (2015-09-29)


### Bug Fixes

* Bug in building of paths ([1c36509](https://github.com/karma-runner/karma-junit-reporter/commit/1c36509)), closes [#65](https://github.com/karma-runner/karma-junit-reporter/issues/65)



<a name="0.3.5"></a>
## [0.3.5](https://github.com/karma-runner/karma-junit-reporter/compare/v0.3.4...v0.3.5) (2015-09-27)


### Bug Fixes

* Add back `outputFile` config ([b4583b5](https://github.com/karma-runner/karma-junit-reporter/commit/b4583b5)), closes [#58](https://github.com/karma-runner/karma-junit-reporter/issues/58)
* Update dependencies ([a1fe8fe](https://github.com/karma-runner/karma-junit-reporter/commit/a1fe8fe))



<a name="0.3.4"></a>
## [0.3.4](https://github.com/karma-runner/karma-junit-reporter/compare/v0.3.3...v0.3.4) (2015-08-28)


### Bug Fixes

* Change how classname is generated ([e4f7ebd](https://github.com/karma-runner/karma-junit-reporter/commit/e4f7ebd))
* **config:** Default `outputDir` to `'.'` ([8fdfb73](https://github.com/karma-runner/karma-junit-reporter/commit/8fdfb73)), closes [#54](https://github.com/karma-runner/karma-junit-reporter/issues/54)

### Features

* Upgrade dependencies to their latest versions ([62b5783](https://github.com/karma-runner/karma-junit-reporter/commit/62b5783))



<a name"0.3.3"></a>
### 0.3.3 (2015-07-23)


#### Bug Fixes

* create browser directory if not exist ([fe14a8d8](https://github.com/karma-runner/karma-junit-reporter/commit/fe14a8d8))


<a name"0.3.2"></a>
### 0.3.2 (2015-07-19)


#### Features

* Allow user to specify filename ([71e0e432](https://github.com/karma-runner/karma-junit-reporter/commit/71e0e432))


<a name"0.3.1"></a>
### 0.3.1 (2015-07-13)


#### Bug Fixes

* handle browser script failure and put the error to browser report ([d5b365ab](https://github.com/karma-runner/karma-junit-reporter/commit/d5b365ab))


<a name"0.3.0"></a>
## 0.3.0 (2015-07-13)


#### Features


test/reporter.spec.js
'use strict'

var chai = require('chai')
var expect = require('chai').expect
var sinon = require('sinon')
var proxyquire = require('proxyquire')
var fs = require('fs')
var libxmljs = require('libxmljs')

// Validation schema is read from a file
var schemaPath = './sonar-unit-tests.xsd'

chai.use(require('sinon-chai'))

function noop () {}

var fakeLogger = {
  create: noop
}

var fakeHelper = {
  normalizeWinPath: noop,
  mkdirIfNotExists: sinon.stub().yields()
}

var fakeFormatError = sinon.spy(function (v) { return v })

var fakeConfig = {
  basePath: __dirname,
  junitReporter: {
    outputFile: ''
  }
}

// Rule of thumb:
// - If you test the new XML format, remember to (within that test) create a new fake reporter,
//   passing it also this line in the fake config: "fakeConfig.junitReporter.xmlVersion: 1"

var fakeBaseReporterDecorator = noop

describe('JUnit reporter', function () {
  var reporterModule
  var reporter

  var fakeFs

  beforeEach(function () {
    fakeFs = {
      writeFile: sinon.spy()
    }

    reporterModule = proxyquire('..', {
      fs: fakeFs
    })
  })

  beforeEach(function () {
    reporter = new reporterModule['reporter:junit'][1](fakeBaseReporterDecorator, fakeConfig, fakeLogger, fakeHelper, fakeFormatError)
  })

  it('should produce valid XML per the new SonarQube reporting format', function () {
    // Two differences in this test, compared to other tests:
    // a) we have a different configuration for the reporter
    // b) need a instantiation of the reporter - the beforeEach doesn't work since it is for old XML
    var fakeBrowser = {
      id: 'Android_4_1_2',
      name: 'Android',
      fullName: 'Android 4.1.2',
      lastResult: {
        error: false,
        total: 1,
        failed: 0,
        netTime: 10 * 1000
      }
    }
    // Static result, since we don't actually produce the result through Karma
    var fakeResult = {
      suite: [
        'Sender',
        'using it',
        'get request'
      ],
      description: 'should not fail',
      log: []
    }
    // Requesting test for NEW xml format. Do not recycle the config used by OTHER tests,
    // since this would ruin them. Remember: since tests can run in undefined order, the side
    // effects (like configuration) must be carefully considered. beforeEach() caters for other tests
    // automatically.
    var newFakeConfig = {
      basePath: __dirname,
      junitReporter: {
        outputFile: '',
        xmlVersion: 1
      }
    }
    // Grab a new reporter, configured with xmlVersion flag
    var nxreporter = new reporterModule['reporter:junit'][1](fakeBaseReporterDecorator, newFakeConfig, fakeLogger, fakeHelper)
    nxreporter.onRunStart([ fakeBrowser ])
    nxreporter.specSuccess(fakeBrowser, fakeResult)
    nxreporter.onBrowserComplete(fakeBrowser)
    nxreporter.onRunComplete()

    var writtenXml = fakeFs.writeFile.firstCall.args[1]

    var xsdString = fs.readFileSync(schemaPath)
    var xsdDoc = libxmljs.parseXml(xsdString)
    var xmlDoc = libxmljs.parseXml(writtenXml)

    xmlDoc.validate(xsdDoc)

    var xsdParseErrorCount = xsdDoc.errors.length
    var xmlParseErrorCount = xmlDoc.errors.length
    var validationErrorCount = xmlDoc.validationErrors.length

    // The 2 tests below are "static", weak tests that find whether a
    // string is present in the XML report
    expect(writtenXml).to.have.string('testCase name="Sender using it get request should not fail"')
    expect(writtenXml).to.have.string('unitTest')
    // The below is the strict, libxml-xsd -based validation result
    expect(validationErrorCount).to.equal(0)
    expect(xsdParseErrorCount).to.equal(0)
    expect(xmlParseErrorCount).to.equal(0)
  })

  it('should include parent suite names in generated test names', function () {
    var fakeBrowser = {
      id: 'Android_4_1_2',
      name: 'Android',
      fullName: 'Android 4.1.2',
      lastResult: {
        error: false,
        total: 1,
        failed: 0,
        netTime: 10 * 1000
      }
    }

    var fakeResult = {
      suite: [
        'Sender',
        'using it',
        'get request'
      ],
      description: 'should not fail',
      log: []
    }

    reporter.onRunStart([ fakeBrowser ])
    reporter.specSuccess(fakeBrowser, fakeResult)
    reporter.onBrowserComplete(fakeBrowser)
    reporter.onRunComplete()

    expect(fakeFs.writeFile).to.have.been.called

    var writtenXml = fakeFs.writeFile.firstCall.args[1]
    expect(writtenXml).to.have.string('testcase name="Sender using it get request should not fail"')
  })

  it('should safely handle special characters', function () {
    var fakeBrowser = {
      id: 'Android_4_1_2',
      name: 'Android',
      fullName: 'Android 4.1.2',
      lastResult: {
        error: false,
        total: 1,
        failed: 1,
        netTime: 10 * 1000
      }
    }

    var fakeResult = {
      suite: [
        'Sender',
        'using it',
        'get request'
      ],
      success: false,
      description: 'should not fail',
      log: ['Expected "👍" to be "👎".']
    }

    reporter.onRunStart([ fakeBrowser ])
    reporter.specSuccess(fakeBrowser, fakeResult)
    reporter.onBrowserComplete(fakeBrowser)
    reporter.onRunComplete()

    expect(fakeFs.writeFile).to.have.been.called

    var writtenXml = fakeFs.writeFile.firstCall.args[1]
    expect(writtenXml).to.have.string('<failure type="">Expected "👍" to be "👎".</failure>')
  })

  it('should safely handle missing suite browser entries when specSuccess fires', function () {
    reporter.onRunStart([])
    // don't try to call null.ele()
    expect(reporter.specSuccess.bind(reporter, { id: 1 }, {})).to.not.throw(TypeError)
  })

  it('should safely handle invalid test result objects when onBrowserComplete fires', function () {
    var badBrowserResult = {
      id: 'Android_4_1_2',
      name: 'Android',
      fullName: 'Android 4.1.2',
      lastResult: {
        error: true,
        netTime: 0
      }
    }

    reporter.onRunStart([ badBrowserResult ])

    // never pass a null value to XMLAttribute via xmlbuilder attr()
    expect(reporter.onBrowserComplete.bind(reporter, badBrowserResult)).not.to.throw(Error)
  })

  it('should safely handle test re-runs triggered by watchers', function () {
    var fakeBrowser = {
      id: 'Android_4_1_2',
      name: 'Android',
      fullName: 'Android 4.1.2',
      lastResult: {
        error: false,
        total: 1,
        failed: 0,
        netTime: 10 * 1000
      }
    }

    reporter.onRunStart([ fakeBrowser ])
    reporter.onBrowserStart(fakeBrowser)

    // When a watcher triggers a second test run, onRunStart() for the second
    // run gets triggered, followed by onRunComplete() from the first test run.
    reporter.onRunStart([ fakeBrowser ])
    reporter.onRunComplete()

    reporter.onBrowserStart(fakeBrowser)
    reporter.onBrowserComplete(fakeBrowser)
    reporter.onRunComplete()
  })
})


gruntfile.js
module.exports = function (grunt) {
  grunt.initConfig({
    pkgFile: 'package.json',
    'npm-contributors': {
      options: {
        commitMessage: 'chore: update contributors'
      }
    },
    conventionalChangelog: {
      release: {
        options: {
          changelogOpts: {
            preset: 'angular'
          }
        },
        src: 'CHANGELOG.md'
      }
    },
    conventionalGithubReleaser: {
      release: {
        options: {
          auth: {
            type: 'oauth',
            token: process.env.GH_TOKEN
          },
          changelogOpts: {
            preset: 'angular',
            releaseCount: 0
          }
        }
      }
    },
    bump: {
      options: {
        pushTo: 'upstream',
        commitFiles: [
          'package.json',
          'CHANGELOG.md'
        ],
        commitMessage: 'chore: release v%VERSION%'
      }
    },
    eslint: {
      target: [
        'index.js',
        'gruntfile.js',
        'test/*.js'
      ]
    },
    simplemocha: {
      options: {
        ui: 'bdd',
        reporter: 'dot'
      },
      unit: {
        src: [
          'test/*.spec.js'
        ]
      }
    }
  })

  require('load-grunt-tasks')(grunt)

  grunt.registerTask('test', ['simplemocha'])
  grunt.registerTask('default', ['eslint', 'test'])

  grunt.registerTask('release', 'Bump the version and publish to NPM.', function (type) {
    grunt.task.run([
      'npm-contributors',
      'bump-only:' + (type || 'patch'),
      'conventionalChangelog',
      'bump-commit',
      'conventionalGithubReleaser',
      'npm-publish'
    ])
  })
}


index.js
var os = require('os')
var path = require('path')
var fs = require('fs')
var builder = require('xmlbuilder')
var pathIsAbsolute = require('path-is-absolute')

/* XML schemas supported by the reporter: 'xmlVersion' in karma.conf.js,
   'XMLconfigValue' as variable here.
   0 = "old", original XML format. For example, SonarQube versions prior to 6.2
   1 = first amended version. Compatible with SonarQube starting from 6.2
*/

// concatenate test suite(s) and test description by default
function defaultNameFormatter (browser, result) {
  return result.suite.join(' ') + ' ' + result.description
}

var JUnitReporter = function (baseReporterDecorator, config, logger, helper, formatError) {
  var log = logger.create('reporter.junit')
  var reporterConfig = config.junitReporter || {}
  // All reporterConfig.something are for reading flags from the Karma config file
  var pkgName = reporterConfig.suite || ''
  var outputDir = reporterConfig.outputDir
  var outputFile = reporterConfig.outputFile
  var useBrowserName = reporterConfig.useBrowserName
  var nameFormatter = reporterConfig.nameFormatter || defaultNameFormatter
  var classNameFormatter = reporterConfig.classNameFormatter
  var properties = reporterConfig.properties
  // The below two variables have to do with adding support for new SonarQube XML format
  var XMLconfigValue = reporterConfig.xmlVersion
  var NEWXML
  var gitlab = reporterConfig.gitlab
  // We need one global variable for the tag <file> to be visible to functions
  var exposee
  var suites = []
  var pendingFileWritings = 0
  var fileWritingFinished = function () {}
  var allMessages = []

  // The NEWXML is just sugar, a flag. Remove it when there are more than 2
  // supported XML output formats.
  if (!XMLconfigValue) {
    XMLconfigValue = 0
    NEWXML = false
  } else {
    // Slack behavior: "If defined, assume to be 1" since we have only two formats now
    XMLconfigValue = 1
    NEWXML = true
  }

  if (outputDir == null) {
    outputDir = '.'
  }

  outputDir = helper.normalizeWinPath(path.resolve(config.basePath, outputDir)) + path.sep

  if (typeof useBrowserName === 'undefined') {
    useBrowserName = true
  }

  baseReporterDecorator(this)

  this.adapters = [
    function (msg) {
      allMessages.push(msg)
    }
  ]

  // Creates the outermost XML element: <unitTest>
  var initializeXmlForBrowser = function (browser) {
    var timestamp = (new Date()).toISOString().substr(0, 19)
    var suite
    if (NEWXML) {
      suite = suites[browser.id] = builder.create('unitTest')
      suite.att('version', '1')
      exposee = suite.ele('file', {'path': 'fixedString'})
    } else {
      if (gitlab === true) {
        var root = builder.create('testsuites')
        suite = suites[browser.id] = root.ele('testsuite')
      } else {
        suite = suites[browser.id] = builder.create('testsuite')
      }
      suite.att('name', browser.name)
      .att('package', pkgName)
      .att('timestamp', timestamp)
      .att('id', 0)
      .att('hostname', os.hostname())
      var propertiesElement = suite.ele('properties')
      propertiesElement.ele('property', {name: 'browser.fullName', value: browser.fullName})

      // add additional properties passed in through the config
      for (var property in properties) {
        if (properties.hasOwnProperty(property)) {
          propertiesElement.ele('property', {name: property, value: properties[property]})
        }
      }
    }
  }

  // This function takes care of writing the XML into a file
  var writeXmlForBrowser = function (browser) {
    // Define the file name using rules
    var safeBrowserName = browser.name.replace(/ /g, '_')
    var newOutputFile
    if (outputFile && pathIsAbsolute(outputFile)) {
      newOutputFile = outputFile
    } else if (outputFile != null) {
      var dir = useBrowserName ? path.join(outputDir, safeBrowserName)
                               : outputDir
      newOutputFile = path.join(dir, outputFile)
    } else if (useBrowserName) {
      newOutputFile = path.join(outputDir, 'TESTS-' + safeBrowserName + '.xml')
    } else {
      newOutputFile = path.join(outputDir, 'TESTS.xml')
    }

    var xmlToOutput = suites[browser.id]

    if (!xmlToOutput) {
      return // don't die if browser didn't start
    }

    pendingFileWritings++
    helper.mkdirIfNotExists(path.dirname(newOutputFile), function () {
      fs.writeFile(newOutputFile, xmlToOutput.end({pretty: true}), function (err) {
        if (err) {
          log.warn('Cannot write JUnit xml\n\t' + err.message)
        } else {
          log.debug('JUnit results written to "%s".', newOutputFile)
        }

        if (!--pendingFileWritings) {
          fileWritingFinished()
        }
      })
    })
  }

  // Return a 'safe' name for test. This will be the name="..." content in XML.
  var getClassName = function (browser, result) {
    var name = ''
    // configuration tells whether to use browser name at all
    if (useBrowserName) {
      name += browser.name
        .replace(/ /g, '_')
        .replace(/\./g, '_') + '.'
    }
    if (pkgName) {
      name += '.'
    }
    if (result.suite && result.suite.length > 0) {
      name += result.suite.join(' ')
    }
    return name
  }

  // "run_start" - a test run is beginning for all browsers
  this.onRunStart = function (browsers) {
    // TODO(vojta): remove once we don't care about Karma 0.10
    browsers.forEach(initializeXmlForBrowser)
  }

  // "browser_start" - a test run is beginning in _this_ browser
  this.onBrowserStart = function (browser) {
    initializeXmlForBrowser(browser)
  }

  // "browser_complete" - a test run has completed in _this_ browser
  // writes the XML to file and releases memory
  this.onBrowserComplete = function (browser) {
    var suite = suites[browser.id]
    var result = browser.lastResult
    if (!suite || !result) {
      return // don't die if browser didn't start
    }

    if (!NEWXML) {
      suite.att('tests', result.total ? result.total : 0)
      suite.att('errors', result.disconnected || result.error ? 1 : 0)
      suite.att('failures', result.failed ? result.failed : 0)
      suite.att('time', (result.netTime || 0) / 1000)
      suite.ele('system-out').dat(allMessages.join() + '\n')
      suite.ele('system-err')
    }

    writeXmlForBrowser(browser)

    // Release memory held by the test suite.
    suites[browser.id] = null
  }

  // "run_complete" - a test run has completed on all browsers
  this.onRunComplete = function () {
    allMessages.length = 0
  }

  // --------------------------------------------
  // | Producing XML for individual testCase    |
  // --------------------------------------------
  this.specSuccess = this.specSkipped = this.specFailure = function (browser, result) {
    var testsuite = suites[browser.id]
    var validMilliTime
    var spec

    if (!testsuite) {
      return
    }

    // New in the XSD schema: only name and duration. classname is obsoleted
    if (NEWXML) {
      if (!result.time || result.time === 0) {
        validMilliTime = 1
      } else {
        validMilliTime = result.time
      }
    }

    // create the tag for a new test case
    /*
    if (NEWXML) {
      spec = testsuite.ele('testCase', {
      name: nameFormatter(browser, result),
      duration: validMilliTime })
    }
    */

    if (NEWXML) {
      spec = exposee.ele('testCase', {
        name: nameFormatter(browser, result),
        duration: validMilliTime })
    } else {
      // old XML format. Code as-was
      spec = testsuite.ele('testcase', {
        name: nameFormatter(browser, result),
        time: ((result.time || 0) / 1000),
        classname: (typeof classNameFormatter === 'function' ? classNameFormatter : getClassName)(browser, result)
      })
    }

    if (result.skipped) {
      spec.ele('skipped')
    }

    if (!result.success) {
      result.log.forEach(function (err) {
        if (!NEWXML) {
          spec.ele('failure', {type: ''}, formatError(err))
        } else {
        // In new XML format, an obligatory 'message' attribute in failure
          spec.ele('failure', {message: formatError(err)})
        }
      })
    }
  }

  // wait for writing all the xml files, before exiting
  this.onExit = function (done) {
    if (pendingFileWritings) {
      fileWritingFinished = done
    } else {
      done()
    }
  }
}

JUnitReporter.$inject = ['baseReporterDecorator', 'config', 'logger', 'helper', 'formatError']

// PUBLISH DI MODULE
module.exports = {
  'reporter:junit': ['type', JUnitReporter]
}
